android/libraries/rib-router-navigator/src/main/kotlin/com/uber/rib/core/RouterAndState.kt

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (C) 2023. Uber Technologies
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.uber.rib.core
+
+import androidx.annotation.VisibleForTesting
+
+/**
+ * Internal class for keeping track of a navigation stack.
+ *
+ * @param state The associated state
+ * @param attachTransition The [RouterNavigator.AttachTransition] associated with this state.
+ * @param detachTransition The [RouterNavigator.DetachTransition] associated with this state.
+ * @param forceRouterCaching Override [RouterNavigatorState.isCacheable] behavior
+ */
+internal class RouterAndState<R : Router<*>, StateT : RouterNavigatorState>(
+  val state: StateT,
+  private val attachTransition: RouterNavigator.AttachTransition<R, StateT>,
+  detachTransition: RouterNavigator.DetachTransition<R, StateT>?,
+  @get:VisibleForTesting val forceRouterCaching: Boolean = false
+) {
+  @get:Synchronized
```

Comment:
Not sure what did you mean by "whole object". I've marked only the `get` and `set` of `_router` by `Synchronized`, this annotation will generate (if not exist) private `$lock` object for locking, if we will have any additional method which doesn't touching `_router`, it will not be synchronized.

Reply:
Also, I can add small internal `synchronizedProperty` delegate, so it will look like : 
`var _router : R? by synchronizedProperty()`

Reply:
Just did some verifications
```kotlin
public class Test {
  @get:Synchronized
  public val a: Int = 3
}
```
Translates to
```java
public final class Test {
   private final int a = 3;

   public final synchronized int getA() {
      return this.a;
   }
}
```
So `@Synchronized` translates to a Java [synchronized method](https://docs.oracle.com/javase/tutorial/essential/concurrency/syncmeth.html`, which synchronizes on the instance of `RouterAndState`.

What I meant with my comment is that if we have another field and mark that other field as `@Synchronized`, the getter for this field would also be synchronized on the class instance, and hence two threads, one accessing `property A` and other accessing `property B`, would have to wait one for the other when not needed. But that's not the case on your impl.

Nevertheless, I see some issues in the code. For example:
```kotlin
  private fun destroyRouterIfNeed() {
    if (!forceRouterCaching && !state.isCacheable()) {
      val routerName = _router?.javaClass?.simpleName
      _router = null
      if (routerName != null) {
        log("Destroying router $routerName was destroyed")
      } else {
        log("Router of ${state.stateName()} state already destroyed")
      }
    }
  }
  ```

This method reads `_router`, then writes to `_router`. But the lock in between those operations is lost, and anything could happen to this field in the meantime. In other words, the whole operation is not atomic, though reading and writing to the property individually are atomic. 

I'd suggest either synchronizing all methods that touch the property for having fully atomic operations, or probably better yet, wrap the property in  [AtomicReference](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.native.concurrent/-atomic-reference/). With the latter, above method could be written as:

```kotlin
private fun destroyRouterIfNeed() {
  if (!forceRouterCaching && !state.isCacheable()) {
    // notice the block can run multiple times, so should be side effect free
    val previous = _router.getAndUpdate { null }
    val routerName = current?.javaClass?.simpleName
    if (routerName != null) {
      log("Destroying router $routerName was destroyed")
    } else {
      log("Router of ${state.stateName()} state already destroyed")
    }
  }
}
```

Reply:
Good point ! Already redesigned for thread safe usage + covered by unit test (test fails without new logic thread-safe)